### ISSUE TITLE
Supporting arbitrary constructors?

### ISSUE THREAD

**Issue by bollwyvl** [goanpeca/setup-miniconda#42]
Thanks for this tool! 

While the name of the project says it all, it seems like the action could install and use any constructor-generated installer: the immediate candidate would be [miniforge](https://github.com/conda-forge/miniforge/releases), and I'm kinda itching to try out their pypy releases for speedier CI (haven't really benchmarked it yet). But I also see the potential for just any-old-constructor, especially as part of a longer pipeline.

The most blunt instrument, just an `installer-url`, could override all the detection, listing, etc. mechanisms. I'm very new to github actions, but would be interested in looking into this, if you are amenable!

**Comment by bollwyvl** [comment]
> The most blunt instrument, just an installer-url

Something along the lines of:
```patch
diff --git a/src/setup.ts b/src/setup.ts
index a6f4b2b..1bab5b8 100644
--- a/src/setup.ts
+++ b/src/setup.ts
@@ -212,6 +212,29 @@ async function downloadMiniconda(
   return { ok: true, data: downloadPath };
 }
 
+async function downloadInstaller(url: string): Promise<Result> {
+  let downloadPath: string;
+
+  // Look for cache to use
+  const cachedInstallerPath = tc.find(url, url);
+
+  if (cachedInstallerPath) {
+    core.info(`Found cache at ${cachedInstallerPath}`);
+    downloadPath = cachedInstallerPath;
+  } else {
+    try {
+      downloadPath = await tc.downloadTool(url);
+      const options = { recursive: true, force: false };
+
+      core.info(`Saving cache...`);
+      await tc.cacheFile(downloadPath, url, url, url);
+    } catch (err) {
+      return { ok: false, error: err };
+    }
+  }
+  return { ok: true, data: downloadPath };
+}
+
 /**
  * Install Miniconda
  *
@@ -542,6 +565,7 @@ async function applyCondaConfiguration(
  * Main conda setup method to handle all configuration options
  */
 async function setupMiniconda(
+  installerUrl: string,
   minicondaVersion: string,
   architecture: string,
   condaVersion: string,
@@ -571,7 +595,13 @@ async function setupMiniconda(
       };
     }
 
-    if (minicondaVersion !== "" || architecture !== "x64") {
+    if (installerUrl !== "") {
+      utils.consoleLog("\n# Downloading Custom Installer...\n");
+      useBundled = false;
+      result = await downloadInstaller(installerUrl);
+      utils.consoleLog("Installing Custom Installer...");
+      result = await installMiniconda(result["data"], useBundled);
+    } else if (minicondaVersion !== "" || architecture !== "x64") {
       utils.consoleLog("\n# Downloading Miniconda...\n");
       useBundled = false;
       result = await downloadMiniconda(3, minicondaVersion, architecture);
@@ -737,6 +767,7 @@ async function setupMiniconda(
  */
 async function run(): Promise<void> {
   try {
+    let installerUrl: string = core.getInput("installer-url");
     let minicondaVersion: string = core.getInput("miniconda-version");
     let condaVersion: string = core.getInput("conda-version");
     let condaBuildVersion: string = core.getInput("conda-build-version");
@@ -773,6 +804,7 @@ async function run(): Promise<void> {
       use_only_tar_bz2: useOnlyTarBz2
     };
     const result = await setupMiniconda(
+      installerUrl,
       minicondaVersion,
       "x64",
       condaVersion,

```

**Comment by goanpeca** [comment]
Hi @bollwyvl sure, that should work fine :-), feel free to open a PR, I can fill in any missing details!

Thanks :-)

**Comment by goanpeca** [comment]
Let me know if you are working on this @bollwyvl and don't forget to start :-p

**Comment by bollwyvl** [comment]
Closed by #43
